fpga_interchange/fasm_generators/xc7.py

```diff
@@ -0,0 +1,549 @@
+            assert m, site_name
+
+            site_loc = int(m.group(1)) % 16
+            site_prefix = "BUFGCTRL.BUFGCTRL_X0Y{}".format(site_loc)
+
+            tile_name = cell_data.tile_name
+
+            self.add_cell_feature((tile_name, site_prefix, "IN_USE"))
+
+            if cell_type == "BUFG":
+                for feature in ["IS_IGNORE1_INVERTED", "ZINV_CE0", "ZINV_S0"]:
+                    self.add_cell_feature((tile_name, site_prefix, feature))
+
+    def handle_slice_routing_bels(self):
+        allowed_routing_bels = list()
+
+        used_muxes = ["SRUSEDMUX", "CEUSEDMUX"]
+        allowed_routing_bels.extend(used_muxes)
+
+        for loc in "ABCD":
+            ff_mux = "{}FFMUX".format(loc)
+            ff5_mux = "{}5FFMUX".format(loc)
+            out_mux = "{}OUTMUX".format(loc)
+            allowed_routing_bels.extend([ff_mux, ff5_mux, out_mux])
+
+        routing_bels = self.get_routing_bels(allowed_routing_bels)
+
+        for site, bel, pin in routing_bels:
+            tile_name, tile_type = self.get_tile_info_at_site(site)
+            slice_prefix = self.get_slice_prefix(site, tile_type)
+
+            if bel in used_muxes:
+                if pin not in ["0", "1"]:
+                    self.add_cell_feature((tile_name, slice_prefix, bel))
+            else:
+                self.add_cell_feature((tile_name, slice_prefix, bel, pin))
+
+    def handle_site_thru(self, site_thru_pips):
+        """
+        This function is currently specialized to add very specific features
+        for pseudo PIPs which need to be enabled to get the correct HW behaviour
+        """
+
+        def get_feature_prefix(site_thru_feature, wire):
+            regex = re.compile(site_thru_feature.regex)
+
+            m = regex.match(wire)
+
+            return site_thru_feature.callback(m) if m else None
+
+        # FIXME: this information needs to be added as an annotation
+        #        to the device resources
+        site_thru_features = list()
+        site_thru_features.append(
+            ExtraFeatures(
+                regex="IOI_OLOGIC([01])_D1",
+                features=["OMUX.D1", "OQUSED", "OSERDES.DATA_RATE_TQ.BUF"],
+                callback=lambda m: "OLOGIC_Y{}".format(m.group(1))))
+        site_thru_features.append(
+            ExtraFeatures(
+                regex="[LR]IOI_ILOGIC([01])_D",
+                features=["ZINV_D"],
+                callback=lambda m: "ILOGIC_Y{}".format(m.group(1))))
+        site_thru_features.append(ExtraFeatures(
+            regex="CLK_HROW_CK_MUX_OUT_([LR])([0-9]+)",
+            features=["IN_USE", "ZINV_CE"],
+            callback=lambda m: "BUFHCE.BUFHCE_X{}Y{}".format(0 if m.group(1) == "L" else 1, m.group(2))
+            ))
+        site_thru_features.append(
+            ExtraFeatures(
+                regex="CLK_BUFG_BUFGCTRL([0-9]+)_I[01]",
+                features=[
+                    "IN_USE", "ZINV_CE0", "ZINV_S0", "IS_IGNORE1_INVERTED"
```

Comment:
I think the features will need to be different depending on whether the I0/I1 route-thru is used (I know this isn't relevant ATM with BUFGCTRL route-thrus disabled but it will be important in the future).

Reply:
I have added a second regex with the other pin used and the corresponding features and added a `TODO` comment to better handle this case, as there is room for generalization I believe.